modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,543 @@
+     * @param procId  If {@code snpName} is {@code null}, is used to find the operation request.
+     * @return Current snapshot checking request by {@code snpName} or {@code procId}.
+     */
+    private @Nullable SnapshotCheckProcessRequest currentRequest(@Nullable String snpName, UUID procId) {
+        return snpName == null
+            ? requests.values().stream().filter(rq -> rq.requestId().equals(procId)).findFirst().orElse(null)
+            : requests.get(snpName);
+    }
+
+    /** Phase 1 beginning: prepare, collect and check local metas. */
+    private IgniteInternalFuture<ArrayList<SnapshotMetadata>> prepareAndCheckMetas(SnapshotCheckProcessRequest extReq) {
+        SnapshotCheckProcessRequest locReq = requests.computeIfAbsent(extReq.snapshotName(), snpName -> extReq);
+
+        if (!locReq.equals(extReq)) {
+            Throwable err = new IllegalStateException("Validation of snapshot '" + extReq.snapshotName()
+                + "' has already started. Request=" + locReq + '.');
+
+            clean(extReq.requestId(), err, null, null);
+
+            return new GridFinishedFuture<>(err);
+        }
+
+        if (!extReq.nodes.contains(kctx.localNodeId())) {
+            if (log.isDebugEnabled()) {
+                log.debug("Skipping snapshot local metadatas collecting for snapshot validation, request=" + extReq
+                    + ". Current node is not required.");
+            }
+
+            return new GridFinishedFuture<>();
+        }
+
+        GridFutureAdapter<ArrayList<SnapshotMetadata>> locMetasChkFut = new GridFutureAdapter<>();
+
+        assert locReq.fut() == null;
+
+        locReq.fut(locMetasChkFut);
+
+        IgniteSnapshotManager snpMgr = kctx.cache().context().snapshotMgr();
+
+        snpMgr.snapshotExecutorService().submit(() -> stopFutureOnAnyFailure(locMetasChkFut, () -> {
+            if (log.isDebugEnabled())
+                log.debug("Checking local snapshot metadatas. Request=" + locReq + '.');
+
+            Collection<Integer> grpIds = F.isEmpty(locReq.groups()) ? null : F.viewReadOnly(locReq.groups(), CU::cacheId);
+
+            // An error can occure when the local future is still null.
+            if (locReq.error() != null)
+                locMetasChkFut.onDone(locReq.error());
+
+            if (locMetasChkFut.isDone())
+                return;
+
+            List<SnapshotMetadata> locMetas = snpMgr.checker().checkLocalMetas(
+                snpMgr.snapshotLocalDir(locReq.snapshotName(), locReq.snapshotPath()),
+                grpIds,
+                kctx.cluster().get().localNode().consistentId()
+            );
+
+            if (!F.isEmpty(locMetas))
+                locReq.meta(locMetas.get(0));
+
+            if (!(locMetas instanceof ArrayList))
+                locMetas = new ArrayList<>(locMetas);
+
+            // A node might have already gone before this. No need to proceed.
+            Throwable locRqErr = locReq.error();
+
+            if (locRqErr != null)
```

Comment:
You already check the error in this thread on L344. I suppose if we check it once and then run the check then we should respond with result of the check